components/encryption/src/manager/mod.rs

```diff
@@ -23,20 +23,22 @@ const KEY_DICT_NAME: &str = "key.dict";
 const FILE_DICT_NAME: &str = "file.dict";
 
 struct Dicts {
-    file_dict: FileDictionary,
-    key_dict: KeyDictionary,
+    file_dict: RwLock<FileDictionary>,
+    key_dict: RwLock<KeyDictionary>,
+    write_lock: Mutex<()>,
     rotation_period: Duration,
     base: PathBuf,
 }
 
 impl Dicts {
     fn new(path: &str, rotation_period: Duration) -> Dicts {
         Dicts {
-            file_dict: FileDictionary::default(),
-            key_dict: KeyDictionary {
+            file_dict: RwLock::new(FileDictionary::default()),
+            key_dict: RwLock::new(KeyDictionary {
```

Comment:
I insist we switch to mutex-based approach instead of rwlock. I did a simple benchmark for concurrent read and write locking. For reads mutex is only faster than rwlock by a small margin, but with rwlock writes can starved.

<img width="336" alt="WeChatWorkScreenshot_ea245b25-b8d3-41d2-950e-73fe61551f78" src="https://user-images.githubusercontent.com/2606959/96388955-617cc980-1161-11eb-9ad1-8c694dc412d4.png">

The test code is here: https://gist.github.com/yiwu-arbug/4f3991e018ddee284c478362ebe1984d

Reply:
btw the test code is modified from @hunterlxt 's: https://gist.github.com/hunterlxt/f56d2cf9a6bae1bb949f4e10f5947aab